gordon_janitor_gcp/plugins/authority.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A GCEAuthority retrieves a list of all instances in all projects that it has
+access to, and which belong to the configured zone. For every project, it will
+create a message containing domain record information and put it into the
+rrset channel. Projects can be filtered by 'project name'. Instances can be
+filtered by tags and metadata.
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    from gordon_janitor_gcp import plugins
+
+    async def run():
+        rrset_channel = asyncio.queue()
+        authority = plugins.get_authority(config, rrset_channel)
+        await authority.start()
+        msg = await rrset_channel.get()
+        print(msg)
+
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(run())
+    # prints: {'zone': 'example.com', 'resourceRecords': [...]}
+"""
+
+import asyncio
+import logging
+
+import aiohttp
+import zope.interface
+from gordon_janitor import interfaces
+
+from gordon_janitor_gcp import exceptions
+from gordon_janitor_gcp.clients import auth
+from gordon_janitor_gcp.clients import gce
+from gordon_janitor_gcp.clients import gcrm
+
+
+class GCEAuthorityBuilder:
+    """Class that helps create a properly configured GCEAuthority object.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+    def __init__(self, config, rrset_channel, **kwargs):
+        self.config = config
+        self.rrset_channel = rrset_channel
+        self.kwargs = kwargs
+        self._loop = asyncio.get_event_loop()
+        self.session = aiohttp.ClientSession(loop=self._loop)
+
+    def _get_or_raise(self, key):
+        item = self.config.get(key)
+        if not item:
+            msg = f'gcp.gce configuration missing "{key}" setting'
+            logging.error(msg)
+            # Close the session if any required config is missing.
+            self.session.close()
```

Comment:
Do we have to create a session before we use it, to avoid needing to close it here?

Reply:
We could let the clients create their own sessions, but that would increase the number of sessions from 1 to 2 here, and to 3 in the internal-facing parts. I'm going for minimizing the number of moving parts - do you think that's too cautious?

Reply:
I'm thinking perhaps to have `self.session` set to `None` on instantiation, and then this:

```py
def build_authority(self):
    self._get_or_raise('zone')
    keyfile_path = self._get_or_raise('keyfile')
    scopes = self._get_or_raise('scopes') 
    self.session = aiohttp.ClientSession()
    crm_client = self.get_crm_client(keyfile_path, scopes)
    # and so on
```